File: src/material/DefaultMaterials.cpp

#include "DefaultMaterials.h"
#include "base/Code.h"
#include "base/RenderMaterial.h"
#include "base/StringHash.h"
#include "base/Renderer.h"
#include <cassert>

namespace three {


  uint32_t MeshBasicMaterial::Type = three::StringHash("MeshBasicMaterial").hash;

  MeshBasicMaterial::MeshBasicMaterial()
  {
    map = 0;
    lightMap = 0;
    envMap = 0;
    offsetRepeat = three::Vector4(0, 0, 1, 1);
    diffuse = three::Color(1, 1, 1, 1);
    opacity = 1;
    flipEnvMap = -1;
    useRefract = 0;
    reflectivity = 1.0;
    refractionRatio = 0.98;
    useTextureMap = false;
    useLightMap = false;
    useEnvMap = false;
    doubleSided = false;
    useVertexColor = false;
    gammaCorrection = false;
    useReflectivity = true;
  }

  void MeshBasicMaterial::apply(Renderer * renderer)
  {
    assert(__renderMaterial && renderer);
    __renderMaterial->uniform(0, offsetRepeat);
    __renderMaterial->uniform(1, diffuse);
    __renderMaterial->uniform(2, opacity);
    __renderMaterial->uniform(3, flipEnvMap);
    __renderMaterial->uniform(4, useRefract);
    __renderMaterial->uniform(5, reflectivity);
    __renderMaterial->uniform(6, refractionRatio);
    renderer->setTexture(map, 0);
    renderer->setTexture(lightMap, 1);
    renderer->setTexture(envMap, 2);
  }

  std::string MeshBasicMaterial::vertexShaderCode() const
  {
    std::vector<std::string> defines;
    if (useTextureMap)
      defines.push_back("USE_MAP");
    if (useLightMap)
      defines.push_back("USE_LIGHTMAP");
    if (useEnvMap)
      defines.push_back("USE_ENVMAP");
    if (doubleSided)
      defines.push_back("DOUBLE_SIDED");
    if (useVertexColor)
      defines.push_back("USE_COLOR");
    if (gammaCorrection)
      defines.push_back("USE_GAMMA");
    if (useReflectivity)
      defines.push_back("USE_REFLECTIVITY");

    return three::Code::generate("basicVertexShader", defines);
  }

  std::string MeshBasicMaterial::fragmentShaderCode() const
  {
    std::vector<std::string> defines;
    if (useTextureMap)
      defines.push_back("USE_MAP");
    if (useLightMap)
      defines.push_back("USE_LIGHTMAP");
    if (useEnvMap)
      defines.push_back("USE_ENVMAP");
    if (doubleSided)
      defines.push_back("DOUBLE_SIDED");
    if (useVertexColor)
      defines.push_back("USE_COLOR");
    if (gammaCorrection)
      defines.push_back("USE_GAMMA");
    if (useReflectivity)
      defines.push_back("USE_REFLECTIVITY");

    return three::Code::generate("basicFragmentShader", defines);
  }

  const char * MeshBasicMaterial::textureName(uint32_t slot) const
  {
    switch (slot)
    {
      case 0:
        return "map";
      case 1:
        return "lightMap";
      case 2:
        return "envMap";
      default:
        return 0;
    }
  }

  const char * MeshBasicMaterial::uniformName(uint32_t slot) const
  {
    switch (slot)
    {
      case 0:
        return "offsetRepeat";
      case 1:
        return "diffuse";
      case 2:
        return "opacity";
      case 3:
        return "flipEnvMap";
      case 4:
        return "useRefract";
      case 5:
        return "reflectivity";
      case 6:
        return "refractionRatio";
      default:
        return 0;
    }
  }

}

File: materials/DefaultMaterials.rb
# Copyright (c) 2012 Daniel Wiberg
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

define 'DefaultMaterials', ['three'] do

  material 'MeshBasicMaterial' do
    vertex_shader :basicVertexShader
    fragment_shader :basicFragmentShader

    option :useTextureMap, :USE_MAP, false
    option :useLightMap, :USE_LIGHTMAP, false
    option :useEnvMap, :USE_ENVMAP, false
    option :doubleSided, :DOUBLE_SIDED, false
    option :useVertexColor, :USE_COLOR, false
    option :gammaCorrection, :USE_GAMMA, false
    option :useReflectivity, :USE_REFLECTIVITY, false

    texture :map
    texture :lightMap
    texture :envMap

    vec4 :offsetRepeat, 0, 0, 1, 1
    color :diffuse, 1, 1, 1, 1
    float :opacity, 1
    float :flipEnvMap, -1
    int :useRefract, 0
    float :reflectivity, 1.0
    float :refractionRatio, 0.98
  end
end